### PR TITLE
make Memory Sanitizer actually work

### PR DESCRIPTION
**Pull request by Shnatsel**
Use -Zbuild-std and -Zsanitizer-memory-track-origins when Memory Sanitizer is requested. This makes MSAN actually useful instead of making it segfault immediately for any target. Rust documentation clearly states that these flags are required: https://doc.rust-lang.org/unstable-book/compiler-flags/sanitizer.html#memorysanitizer

Tested on `image-png` fuzz target. It segfaulted on startup prior to this change with `-s memory` enabled and works fine now.

This may still fail in presence of C code linked into the binary, but it's a start.

### CODE REVIEW

**Comment by alex** [comment]
I can't remember what we have in the way of automated tests for cargo-fuzz, is it practical to add one for this?

**Comment by Shnatsel** [comment]
I've added tests. They pass locally but fail on CI because CI environment does not have the `rust-src` component installed, which is required for `-Z build-std`

**Comment by fitzgen** [comment]
> I've added tests. They pass locally but fail on CI because CI environment does not have the `rust-src` component installed, which is required for `-Z build-std`

Can you modify CI to install the `rust-src` component, so that the tests pass? Should be a one line `rustup` invocation here: https://github.com/rust-fuzz/cargo-fuzz/blob/master/.github/workflows/main.yml#L18

**Comment by Shnatsel** [comment]
Done. CI now passes.

File: tests/tests/main.rs [hunk 1]
```diff
@@ -295,6 +295,72 @@ fn run_without_sanitizer_with_crash() {
...
+            "msan_no_crash",
+            r#"
+                #![no_main]
+                use libfuzzer_sys::fuzz_target;
+
+                fuzz_target!(|data: &[u8]| {
+                    let test_data = vec![0,1,2,3];
+                    let accessed_value = test_data[0];
+                    // prevent the read from being optimized out
+                    println!("{}", accessed_value);
```
[the remark this answers is not in the document]
Comment (Shnatsel) [reply tests/tests/main.rs]: Good call. Done.

**Comment by Shnatsel** [comment]
Thanks! When a new release with this and other recent changes can be expected?

Also, should I make some kind of announcement, perhaps on reddit? There was a lot of noteworthy changes recently (working msan, faster defaults, faster release mode due to single codegen unit) and even more compared to the beginning of the year ("no sanitizers" mode, more speedups)

**Comment by fitzgen** [comment]
I can cut a new release today.

Sure, an announcement post on r/rust and users.rust-lang.org sounds good :+1:

**Comment by fitzgen** [comment]
https://crates.io/crates/cargo-fuzz/0.8.0 published!

**Comment by Shnatsel** [comment]
Thanks!

How does this sound for an announcement? https://hackmd.io/@xmxSgN6MSImJPcY4MOKfjA/rybqU8GA8

**Comment by fitzgen** [comment]
Sounds great! Since these things span multiple releases, it might be worth mentioning the versions spanned.

**Comment by fitzgen** [comment]
Maybe worth linking to the changelog too: https://github.com/rust-fuzz/cargo-fuzz/blob/master/CHANGELOG.md

**Comment by Shnatsel** [comment]
Posted: 

https://www.reddit.com/r/rust/comments/hg6x3s/cargofuzz_is_now_10x_faster_better_supports/

https://users.rust-lang.org/t/cargo-fuzz-is-now-10x-faster-better-supports-sanitizers/44957

**Comment by fitzgen** [comment]
Great, thanks! And thanks again for your recent contributions!

**Pull request merged**